substrate/frame/transaction-payment/asset-conversion-tx-payment/src/payment.rs

```diff
@@ -49,150 +53,219 @@ pub trait OnChargeAssetTransaction<T: Config> {
 		asset_id: Self::AssetId,
-		initial_asset_consumed: AssetBalanceOf<T>,
-	) -> Result<AssetBalanceOf<T>, TransactionValidityError> {
-		// Refund the native asset to the account that paid the fees (`who`).
-		// The `who` account will receive the "fee_paid - corrected_fee" refund.
-		<T::OnChargeTransaction>::correct_and_deposit_fee(
-			who,
-			dispatch_info,
-			post_info,
-			corrected_fee,
-			tip,
-			fee_paid,
-		)?;
-
-		// calculate the refund in native asset, to swap back to the desired `asset_id`
-		let swap_back = received_exchanged.saturating_sub(corrected_fee);
-		let mut asset_refund = Zero::zero();
-		if !swap_back.is_zero() {
-			// If this fails, the account might have dropped below the existential balance or there
-			// is not enough liquidity left in the pool. In that case we don't throw an error and
-			// the account will keep the native currency.
-			match CON::swap_exact_tokens_for_tokens(
-				who.clone(), // we already deposited the native to `who`
-				vec![
-					N::get(),        // we provide the native
-					asset_id.into(), // we want asset_id back
-				],
-				swap_back,   // amount of the native asset to convert to `asset_id`
-				None,        // no minimum amount back
-				who.clone(), // we will refund to `who`
-				false,       // no need to keep alive
+		already_withdrawn: Self::LiquidityInfo,
+	) -> Result<BalanceOf<T>, TransactionValidityError> {
+		let (fee_paid, initial_asset_consumed) = already_withdrawn;
+		let refund_amount = fee_paid.peek().saturating_sub(corrected_fee);
+		let (fee_in_asset, adjusted_paid) = if refund_amount.is_zero() ||
+			F::total_balance(asset_id.clone(), who).is_zero()
+		{
+			// Nothing to refund or the account was removed be the dispatched function.
+			(initial_asset_consumed, fee_paid)
+		} else if asset_id == A::get() {
+			// The `asset_id` is the target asset, we do not need to swap.
+			let (refund, fee_paid) = fee_paid.split(refund_amount);
+			if let Err(refund) = F::resolve(who, refund) {
+				let fee_paid = fee_paid.merge(refund).map_err(|_| {
+					// The error should never occur since `fee_paid` and `refund` are
+					// credits of the same asset.
+					InvalidTransaction::Payment
+				})?;
+				(initial_asset_consumed, fee_paid)
+			} else {
+				(fee_paid.peek().saturating_sub(refund_amount), fee_paid)
+			}
+		} else {
+			// Check if the refund amount can be swapped back into the asset used by `who` for fee
+			// payment.
+			let refund_asset_amount = S::quote_price_exact_tokens_for_tokens(
+				A::get(),
+				asset_id.clone(),
+				refund_amount,
+				true,
 			)
-			.ok()
-			{
-				Some(acquired) => {
-					asset_refund = acquired
-						.try_into()
-						.map_err(|_| TransactionValidityError::from(InvalidTransaction::Payment))?;
-				},
-				None => {
-					Pallet::<T>::deposit_event(Event::<T>::AssetRefundFailed {
-						native_amount_kept: swap_back,
-					});
-				},
+			// No refund given if it cannot be swapped back.
+			.unwrap_or(Zero::zero());
```

Comment:
Can it have an early return here?

Reply:
I think I cannot without an additional arm. But I add a condition below, to not do an attempt for zero balance deposit, and one statement below we return.